src/core/infer_request.cc

```diff
@@ -0,0 +1,456 @@
+    const uint64_t batch_byte_size, const InferSharedMemory& shared_memory,
+    InferenceRequest::Input** input)
+{
+  std::vector<int64_t> lshape;
+  for (const auto d : shape) {
+    lshape.push_back(d);
+  }
+
+  const auto& pr = inputs_.emplace(std::make_pair(
+      name,
+      InferenceRequest::Input(name, lshape, batch_byte_size, shared_memory)));
+  if (!pr.second) {
+    return Status(
+        RequestStatusCode::INVALID_ARG, "input '" + name + "' already exists");
+  }
+
+  if (input != nullptr) {
+    *input = std::addressof(pr.first->second);
+  }
+
+  return Status::Success;
+}
+
+Status
+InferenceRequest::AddInput(
+    const std::string& name, const std::vector<int64_t>& shape,
+    const uint64_t batch_byte_size, const InferSharedMemory& shared_memory,
+    InferenceRequest::Input** input)
+{
+  const auto& pr = inputs_.emplace(std::make_pair(
+      name,
+      InferenceRequest::Input(name, shape, batch_byte_size, shared_memory)));
+  if (!pr.second) {
+    return Status(
+        RequestStatusCode::INVALID_ARG, "input '" + name + "' already exists");
+  }
+
+  if (input != nullptr) {
+    *input = std::addressof(pr.first->second);
+  }
+
+  return Status::Success;
+}
+
+Status
+InferenceRequest::RequestOutput(
+    const std::string& name, const uint32_t classification_cnt)
+{
+  const auto& pr = requested_outputs_.emplace(std::make_pair(
+      name, InferenceRequest::RequestedOutput(name, classification_cnt)));
+  if (!pr.second) {
+    return Status(
+        RequestStatusCode::INVALID_ARG,
+        "output '" + name + "' already requested");
+  }
+
+  return Status::Success;
+}
+
+Status
+InferenceRequest::RequestOutput(
+    const std::string& name, const uint32_t classification_cnt,
+    const InferSharedMemory& shared_memory)
+{
+  const auto& pr = requested_outputs_.emplace(std::make_pair(
+      name, InferenceRequest::RequestedOutput(
+                name, classification_cnt, shared_memory)));
+  if (!pr.second) {
+    return Status(
+        RequestStatusCode::INVALID_ARG,
+        "output '" + name + "' already requested");
+  }
+
+  return Status::Success;
+}
+
+Status
+InferenceRequest::AppendInputData(
+    const char* name, const void* base, size_t byte_size,
+    TRTSERVER_Memory_Type memory_type, int64_t memory_type_id)
+{
+  auto pr = input_map_.insert({name, nullptr});
+  std::shared_ptr<Memory>& smem = pr.first->second;
+  if (pr.second) {
+    smem.reset(new MemoryReference());
+  }
+
+  if (byte_size > 0) {
+    std::static_pointer_cast<MemoryReference>(smem)->AddBuffer(
+        static_cast<const char*>(base), byte_size, memory_type, memory_type_id);
+  }
+
+  return Status::Success;
+}
+
+Status
+InferenceRequest::SetInputData(
+    const std::string& name, const std::shared_ptr<Memory>& data)
+{
+  auto pr = input_map_.insert({name, data});
+  if (!pr.second) {
+    return Status(
+        RequestStatusCode::INVALID_ARG,
+        "input '" + name + "' already has data, can't overwrite");
```

Comment:
We may want some override method in the future to reuse a request object

Reply:
Yes, this change doesn't allow reuse. That will be a new TRTSERVER API that exposes InferenceRequest with more methods to modify it in place.